### PR TITLE
Fix for issue 2696: connection flow-control double-counting from zero-length STREAM frames

### PR DESCRIPTION
**Pull request by jbaldassari**
Fix for https://github.com/cloudflare/quiche/issues/2696

A zero-length non-`FIN` `STREAM` frame advances the peer's largest received offset (RFC 9000 Section 19.8). The receive path charges connection flow control from that offset, but `RecvBuf::write()` discards empty non-`FIN` buffers without recording the new high-water mark. When in-flight data covering the gap later arrives, the gap is charged a second time.

`send_single()` can produce such frames when the remaining packet capacity fits the encoded `STREAM` header but no payload. This occurs at large stream IDs and offsets, where the actual header exceeds `MAX_STREAM_OVERHEAD`. The bytes preceding the frame's offset have already been charged to the sender's `tx_data`, so the frame is legal but redundant and exposes the receiver accounting bug.

Between two quiche endpoints, the inflation accumulates silently until a legal frame appears to exceed `MAX_DATA`. The receiver then closes the connection with `FLOW_CONTROL_ERROR`, even though the sender has consumed exactly the credit it was granted. In long-running bulk transfers over a lossy path, this caused 1–3 spurious connection closures per hour. Instrumented captures showed `rx_data` exceeding the sum of the stream high-water marks by exact whole-frame multiples.

This change:
1. Advances the `RecvBuf` high-water mark before returning for an empty non-`FIN` buffer, keeping the connection flow-control charge and stream state in lockstep.
2. Defensively suppresses zero-length non-`FIN` `STREAM` frames in `send_single()`. When packet capacity prevents stream data from being emitted, the sender remains marked as size-limited rather than app-limited.

The `empty_stream_frame` expectations are updated accordingly. The previous sequence also depended on accepting a final size below the largest received offset, which RFC 9000 Section 4.5 forbids.

New regression coverage:
* Injects an empty forward `STREAM` frame and verifies that its offset is charged exactly once. Without the receiver fix, this fails with a spurious `Error::FlowControl`.
* Exercises the sender at a large stream offset across the header-only packet-capacity window, verifies that no zero-length non-`FIN` frame is emitted, and confirms that the queued data is subsequently delivered intact.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ghedo any chance we can get this one reviewed and merged ? People do hit this when using netty.

File: quiche/src/stream/recv_buf.rs [hunk 1]
```diff
@@ -121,8 +121,16 @@ impl RecvBuf {
...
+        // No need to store empty buffer that doesn't carry the fin flag, but
+        // its offset still advances the largest received offset (RFC 9000
+        // Section 19.8). The caller charges connection-level flow control
+        // from that offset, so the high-water mark must advance in lockstep:
+        // otherwise the gap up to the frame's offset is charged a second
+        // time when the in-flight data covering it arrives, and the
+        // accumulated double-counting eventually makes a perfectly legal
+        // frame appear to exceed MAX_DATA.
         if !buf.fin() && buf.is_empty() {
+            self.len = cmp::max(self.len, buf.max_off());
```
Comment (ghedo) [inline quiche/src/stream/recv_buf.rs:128]: This early return breaks the drained-stream invariant that `off == len`. `Connection` already marks `max_off_delta` as consumed for a draining STREAM frame, so a later `RESET_STREAM` credits `final_size - self.off` again. Could we update `off` when `self.drain` is true and add a regression test covering shutdown, an empty non-FIN STREAM frame, then `RESET_STREAM`?

**Comment by ghedo** [comment]
Also, the comments are quite wordy and could probably be made more concise.

**Comment by jbaldassari** [comment]
Thanks for taking a look, @ghedo .  I appreciate the feedback as I'm still new to this code.  I believe I've addressed your comments, but please let me know if there is anything else you'd like me to change.

**Review by ghedo — APPROVED** [review]

**Review by inamdarminaz — APPROVED** [review]
